appinventor/components/src/com/google/appinventor/components/runtime/Form.java

```diff
@@ -695,8 +696,29 @@ void ReplayFormOrientation() {
     }
   }
 
+  private Integer generateHashCode(AndroidViewComponent component, PercentStorageRecord.Dim dim) {
+    if (dim == PercentStorageRecord.Dim.HEIGHT) {
+      return component.hashCode() * 2 + 1;
+    } else {
+      return component.hashCode() * 2;
+    }
+  }
+
   public void registerPercentLength(AndroidViewComponent component, int length, PercentStorageRecord.Dim dim) {
-    dimChanges.add(new PercentStorageRecord(component, length, dim));
+    // dimChanges.add(new PercentStorageRecord(component, length, dim));
+    PercentStorageRecord r = new PercentStorageRecord(component, length, dim);
+    Integer key = generateHashCode(component, dim);
+    dimChanges.put(key, r);
+  }
+
+  public void unregisterPercentLength(AndroidViewComponent component, PercentStorageRecord.Dim dim) {
+    // iterate map, remove all entry match this
+    for (Integer key : dimChanges.keySet()) {
```

Comment:
This code can be simplified to `dimChanges.remove(generateHashCode(component, dim));` There will be at most one thing in the map, and removing a key that isn't in the map is a no-op. It's also faster because the HashMap will provide O(1) amortized performance versus iterating over every single entry.